azkaban-common/src/main/java/azkaban/executor/container/watch/FlowStatusUpdatingListener.java

```diff
@@ -0,0 +1,317 @@
+  private final ConcurrentMap<String, AzPodStatusMetadata> podStatusMap;
+  private final AtomicLong flowContainerEventCount = new AtomicLong(0);
+
+  @Inject
+  public FlowStatusUpdatingListener(Props azkProps,
+      ContainerizedImpl containerizedImpl,
+      ExecutorLoader executorLoader,
+      AlerterHolder alerterHolder) {
+    requireNonNull(azkProps, "azkaban properties must not be null");
+    requireNonNull(containerizedImpl, "container implementation must not be null");
+    requireNonNull(executorLoader, "executor loader must not be null");
+    requireNonNull(alerterHolder, "alerter holder must not be null");
+    this.containerizedImpl = containerizedImpl;
+    this.executorLoader = executorLoader;
+    this.alerterHolder = alerterHolder;
+
+    maxCacheEntries =
+        azkProps.getInt(ContainerizedDispatchManagerProperties.KUBERNETES_WATCH_DRIVER_THREAD_POOL_SIZE,
+        DEFAULT_EVENT_CACHE_MAX_ENTRIES);
+    this.podStatusCache = CacheBuilder.newBuilder()
+        .maximumSize(maxCacheEntries)
+        .recordStats()
+        .build();
+    this.podStatusMap = podStatusCache.asMap();
+  }
+
+  @VisibleForTesting
+  public ContainerizedImpl getContainerizedImpl() {
+    return this.containerizedImpl;
+  }
+
+  @VisibleForTesting
+  public ExecutorLoader getExecutorLoader() {
+    return this.executorLoader;
+  }
+
+  /**
+   * Validate the the transition from the current state to the new state in the {@code event}
+   * is supported. Note that the validation is on best-effort basis as the cache may not always
+   * contain an event for all running flows. This could happen, for example, during webserver
+   * restarts.
+   * @param event pod watch event
+   */
+  private void validateTransition(AzPodStatusMetadata event) {
+    AzPodStatus currentStatus = AzPodStatus.AZ_POD_UNSET;
+    if (podStatusMap.containsKey(event.getPodName())) {
+      currentStatus = podStatusMap.get(event.getPodName()).getAzPodStatus();
+    }
+    logger.debug(format("Transition requested from %s -> %s, for pod %s",
+        currentStatus,
+        event.getAzPodStatus(),
+        event.getPodName()));
+
+    if (!AzPodStatusUtils.isTransitionValid(currentStatus, event.getAzPodStatus())) {
+      IllegalStateException ise = new IllegalStateException(
+          format("Pod status transition is not supported %s -> %s, for pod %s",
+              currentStatus,
+              event.getAzPodStatus(),
+              event.getPodName()));
+      logger.error("Unsupported state transition.", ise);
+      throw ise;
+    }
+  }
+
+  /**
+   * Includes any validations needed on event that are common to all {@link AzPodStatus}
+   *
+   * @param event pod watch event
+   */
+  private void validateAndPreProcess(AzPodStatusMetadata event) {
+    // Confirm that event is indeed a for a FlowContainer and has the corresponding metadata
+    if (!event.getFlowPodMetadata().isPresent()) {
+      IllegalStateException ise = new IllegalStateException(
```

Comment:
Just wondering shall we use Watch exception here?

Reply:
I thought about it as well but the current exception seemed a lot more direct of a match, which in general is the recommended practice as well. Let me know if there is still a concern about this.

Reply:
No concern. Sounds good.